shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+ * This class assumes, the data is Columnar and the plugin is on
+ */
+class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+  override def supportsColumnarInput(schema: Seq[Attribute]): Boolean = true
+
+  override def supportsColumnarOutput(schema: StructType): Boolean = true
+
+  /**
+   * Convert an `RDD[ColumnarBatch]` into an `RDD[CachedBatch]` in preparation for caching the data.
+   * This method uses Parquet Writer on the GPU to write the cached batch
+   * @param input the input `RDD` to be converted.
+   * @param schema the schema of the data being stored.
+   * @param storageLevel where the data will be stored.
+   * @param conf the config for the query.
+   * @return The data converted into a format more suitable for caching.
+   */
+  override def convertColumnarBatchToCachedBatch(input: RDD[ColumnarBatch],
+                                       schema: Seq[Attribute],
+                                       storageLevel: StorageLevel,
+                                       conf: SQLConf): RDD[CachedBatch] = {
+    val parquetCB = input.map(batch => {
+      var gpuCB: ColumnarBatch = null
+      try {
+        // check if the data is already on GPU
+        if (!batch.column(0).isInstanceOf[GpuColumnVector]) {
+          val s = StructType(
+            schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
+          gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
+          batch.close()
+        } else {
+          gpuCB = batch
+        }
+        // now compress it using ParquetWriter
+        compressColumnarBatchWithParquet(gpuCB)
+      } finally {
+        gpuCB.close()
+      }
+    })
+    parquetCB
+  }
+
+  private def compressColumnarBatchWithParquet(gpuCB: ColumnarBatch) = {
+    val buffer = new ParquetCachedBatch(gpuCB.numRows())
+    withResource(GpuColumnVector.from(gpuCB)) { table =>
+      withResource(Table.writeParquetChunked(ParquetWriterOptions.DEFAULT, buffer)) { writer =>
+        writer.write(table)
+      }
+    }
+    buffer.writeBuffers()
+    buffer.asInstanceOf[CachedBatch]
+  }
+
+  /**
+   * This method decodes the CachedBatch leaving it on the GPU to avoid the extra copying back to
+   * the host
+   * @param input
+   * @param cacheAttributes
+   * @param selectedAttributes
+   * @param conf
+   * @return
+   */
+  def gpuConvertCachedBatchToColumnarBatch(input: RDD[CachedBatch],
+    cacheAttributes: Seq[Attribute],
+    selectedAttributes: Seq[Attribute],
+    conf: SQLConf): RDD[ColumnarBatch] = {
+    val cbRdd = convertCachedBatchToColumnarInternal(input, cacheAttributes, selectedAttributes)
+    cbRdd
+  }
+
+  private def convertCachedBatchToColumnarInternal(input: RDD[CachedBatch],
+    cacheAttributes: Seq[Attribute],
+    selectedAttributes: Seq[Attribute]) = {
+    val requestedColumnIndices = selectedAttributes.map(a =>
+      cacheAttributes.map(_.exprId).indexOf(a.exprId))
+
+    // if plugin enabled
```

Comment:
I don't think this is 100% true any more because it gets called from multiple places.